### ISSUE TITLE
Empty scene with "Create Root Node" widget adds extra margin to SceneTree

### ISSUE THREAD

**Issue by akien-mga** [godotengine/godot#25281]
**Godot version:**
Current master (7b4e7d56)

**OS/device including version:**
Mageia 7 Linux x86_64

**Issue description:**
When switching from a saved scene (with nodes in the SceneTree) to an empty new scene with the Create Root Node widget, the width of the SceneTree dock changes. The Create Root Node widget seems to add some extra margin.

It doesn't seem to be dependent on the existing width of the dock, even if I make it much wider than needed to show its contents, the empty scene will expand it.

![screenshot_20190124_102928](https://user-images.githubusercontent.com/4701338/51668902-8b707e80-1fc3-11e9-84f9-0a776218c1ec.png)
![screenshot_20190124_103407](https://user-images.githubusercontent.com/4701338/51668919-962b1380-1fc3-11e9-841b-b0c937645eaf.png)

Tested with French locale, but the same happens in English.

**Steps to reproduce:**
- Create a scene, add a root node, save it
- Create a scene, don't add a root node
- Switch between both scenes and see the SceneTree dock expand and contract

**Comment by yehjaemin** [comment]
I was able to reproduce the bug on macOS Mojave 10.14.2.

I'd really like to solve this, but I'm unfamiliar with the structure of the editor source code. If it's not too much to ask, can someone point me to the files that have to do with the dock or the editor ui? At the moment, I'm trying to go through all the files in godot/editor.

**Comment by ankitpriyarup** [comment]
Even switching between Scene & Import tab on a new scene reproduce this issue. I'm trying to solve it too.

**Comment by azagaya** [comment]
And have you tries setting set_clip_text to true y all the buttons in the empty scene tree dock?

**Comment by ankitpriyarup** [comment]
> And have you tried setting set_clip_text to true y all the buttons in the empty scene tree dock?

Unfortunately, It still doesn't work

**Comment by azagaya** [comment]
That's strange.. it works in my pc with this:
```
	Button *button_2d = memnew(Button);
	beginner_node_shortcuts->add_child(button_2d);
	button_2d->set_text(TTR("2D Scene"));
        button_2d->set_clip_text(true);
	button_2d->set_icon(get_icon("Node2D", "EditorIcons"));
	button_2d->connect("pressed", this, "_tool_selected", make_binds(TOOL_CREATE_2D_SCENE, false));

	Button *button_3d = memnew(Button);
	beginner_node_shortcuts->add_child(button_3d);
	button_3d->set_text(TTR("3D Scene"));
        button_3d->set_clip_text(true);
	button_3d->set_icon(get_icon("Spatial", "EditorIcons"));
	button_3d->connect("pressed", this, "_tool_selected", make_binds(TOOL_CREATE_3D_SCENE, false));

	Button *button_ui = memnew(Button);
	beginner_node_shortcuts->add_child(button_ui);
        button_ui->set_text(TTR("User Interface"));
        button_ui->set_clip_text(true);
	button_ui->set_icon(get_icon("Control", "EditorIcons"));
        button_ui->connect("pressed", this, "_tool_selected", make_binds(TOOL_CREATE_USER_INTERFACE, false));

        VBoxContainer *favorite_node_shortcuts = memnew(VBoxContainer);
        favorite_node_shortcuts->set_name("FavoriteNodeShortcuts");
        node_shortcuts->add_child(favorite_node_shortcuts);

	Button *button_custom = memnew(Button);
        node_shortcuts->add_child(button_custom);
        button_custom->set_text(TTR("Custom Node"));
        button_custom->set_clip_text(true);
        button_custom->set_icon(get_icon("Add", "EditorIcons"));
	button_custom->connect("pressed", this, "_tool_selected", make_binds(TOOL_NEW, false));
```

**Comment by azagaya** [comment]
I

> Also, I noticed that changing text within the HBoxContainer affects the issue. Reducing it to like 'Create Root' fixes the issue while changing the text to something big makes the issue even worse.

Yes, i've noticed that too, and i asked about that in the irc channel. That's why i tried with clip in the first place... in my opinion is something wrong with the resizing of buttons. I should not be directly related with string length because, as @akien-mga pointed outin the irc, making the dock bigger does not solve the issue.. so i think is the resizing of button...

**Comment by azagaya** [comment]
Here i have it working with the clip workarraound

![vokoscreen20190125_090457](https://user-images.githubusercontent.com/46932830/51745138-ba005f00-2080-11e9-918f-5e15b49ff65d.gif)

**Comment by azagaya** [comment]
Hi,
I found that commenting the line 797 (`ms.x = MAX(ms.x, cms.x);`) of tab_container.cpp also seems to solve this. Could anyone confirm this? the issue may be related with geting combined minimum size if so.

**Comment by ankitpriyarup** [comment]
Yeah! That solves the issue here too.

**Comment by azagaya** [comment]
So we should continue investigating in that path, i think.

**Comment by azagaya** [comment]
In tab_container, the get_minimum_size function only takes into account the one that are visible. Taking away that (i mean, without skipping non visible tabls) also seems to help.

**Comment by azagaya** [comment]
Hi,
I'm not sure neither why are they those lines, because as is only works for visible tabs, ms will start being zero, won't it? So taking out those lines is the same than just setting minimum x size to 0 plus what the stylebox gives you at the end... that's why i don't think this is the intended way.. i think is better if we ommit the lines 794 and 795, so tab container get its size according to every tab (not just visibles) and then line 797 and 798 gets sense... but this does not explain the cause of the error... just make the symptoms dissappear.

**Comment by yehjaemin** [comment]
This might be a far reach, but maybe it only accounts for visible tabs in the case where there are many tabs open, and accounting for all tabs, invisible and visible, would begin to affect performance? It seems really unlikely, though.

As for the get_combined_minimum_size method,
```
Size2 Control::get_combined_minimum_size() const {

	if (!data.minimum_size_valid) {
		const_cast<Control *>(this)->_update_minimum_size_cache();
	}
	return data.minimum_size_cache;
}
```
it seems to only return the private variable minimum_size_cache. In conjunction with the update_minimum_size_cache method
```
void Control::_update_minimum_size_cache() {

	Size2 minsize = get_minimum_size();
	minsize.x = MAX(minsize.x, data.custom_minimum_size.x);
	minsize.y = MAX(minsize.y, data.custom_minimum_size.y);

	bool size_changed = false;
	if (data.minimum_size_cache != minsize)
		size_changed = true;

	data.minimum_size_cache = minsize;
	data.minimum_size_valid = true;

	if (size_changed)
		minimum_size_changed();
}
```
I'm unsure if the bug originates from these methods.

**Comment by azagaya** [comment]
@yehjaemin can you see if removing line  `ms.x = MAX(ms.x, cms.x);` also fixes this https://github.com/godotengine/godot/issues/21404 ? because the other approach does "fix" it..

**Comment by yehjaemin** [comment]
Sorry, I'm not really sure what you mean - I already confirmed that commenting the line out works

**Comment by azagaya** [comment]
Don't worry.. i meant if the fixes proposed here also fix another issue that is open.. i'll test tomorrow.

**Comment by ankitpriyarup** [comment]
get_minimum_size() function of TabContainer seems broken removing looks logically correct to me since then it's base class Control get_minimum_size() function will be called. It does not fix #21404  but it still solves this issue as well as #25317 
Maybe we should make a pull request and wait for another senior developer to validate

**Comment by azagaya** [comment]
I like more using invisible tabs to also calculate the minimum size inside the loop..  it also fixes the bug with proyect export wibdow

**Comment by azagaya** [comment]
ankitpriyarup: i've seen your pull request.. however, i must say i don't think is a good idea to return ms.x and ms.y as zero when tabs are less or equal than 2. You know, as you can change tabs from one dock to another, this won't work if i decide to move some tab to another dock and making then the amount of tabs more than 2, leaving us with the same issue again. This would only work if you just leave two tabs as default in any dock.

**Comment by ankitpriyarup** [comment]
You are right, just observed that

**Comment by azagaya** [comment]
Let me dig a little more, if not i will do a PR with the fix of using invisible tabs also in the get_minimum_size function... it works well

**Comment by azagaya** [comment]
I think it also gets solved with my solution.. let me try

**Comment by azagaya** [comment]
My solution also fixes #25346... I think there is something else to solve with tabs resize, but this may work..
